Feature/monitor

NathanealV opened the pull request:
- [Implemented EmbedBuilder and Pagination](https://github.com/ib-ai/ib.py/commit/bbcb1042a08f38125e4c30f373b5e2b2229b1cac)
- [Implemented is_moderator check](https://github.com/ib-ai/ib.py/commit/d2039ad9847c530ceb4e5289cd8bb21139512e54)

- [Abstracted available_subcommands and updated implementation in `updates`](https://github.com/ib-ai/ib.py/commit/0d1a11d579d2a08ea5b3d3c8ef87e4c36ed452be) 
- [Added monitor commands and RegexConverter](https://github.com/ib-ai/ib.py/commit/d863e2fa58ae9f51dcf9b8f154dd516a2f25e679)

raynichc commented on utils/checks.py:
```diff
@@ -1,13 +1,15 @@
+    assert await commands.guild_only().predicate(ctx)
+    guild_data = await GuildData.get_or_none(guild_id=ctx.guild.id)
+    moderator_role_id = guild_data.moderator_id if guild_data else None
+    return any(role.id == moderator_role_id for role in ctx.author.roles) if moderator_role_id else False
```
Is the ternary function here explicitly needed? I will in theory be faster, if there is no moderator role, however, that is essentially never. And if there is no moderator role, it'll just be None, but then the any check will always fail, as none of the role ids will be None.

NathanealV replied:
It was originally added so that we skip going through all the roles and just return early if there is no moderator role (so yeah a performance related addition). So no, it's not explicitly needed and will continue to return false if the moderator role is `None` (was considered before). Can remove it 👍

NathanealV replied:
> Actually, I suggested making the `moderator_role_id == None` case distinct for the sake of error messaging; I have this in a local branch:
> 
> ```python
> class GuildDataMissing(commands.CommandError):
>     pass
> ...
> async def _is_moderator(ctx: commands.Context):
>     ...
>     if not moderator_role_id:
>         raise GuildDataMissing('Moderator role not set for this guild.')
>     ...
> ```
> 
> It's more descriptive than discord.py's "The check functions for command update failed.", but I am not sure how necessary it is.

For this reason, I'm going to keep this in the codebase, which can be changed later when cog checks are changed again. How does this sound @raynichc ?

raynichc replied:
I think that should be okay. Slight nitpick with the comment though, it should be `moderator_role_id is None`, though in the code I think the `not moderator_role_id` is good.

ilikecubesnstuff commented on utils/pagination.py:
```diff
@@ -0,0 +1,63 @@
+        await self.update_page(interaction)
+    
+    async def update_page(self, interaction: discord.Interaction):
+        if self.current_page == 0:
```
It might be more concise to have
```py
for i in self.children: i.disabled = False
if self.current_page == 0:
    self.children[0].disabled = True
    self.children[1].disabled = True
if self.current_page == len(self.embeds) - 1:
    self.children[-1].disabled = True
    self.children[-2].disabled = True
```
I would isolate this logic into something like `async def update_buttons()`, and then call it as well as `await interaction.response.edit_message()` within each button's method. `update_buttons` could be called inside `return_paginated_embed_view` too, since it would disable all buttons when `self.current_page == 0 == len(self.embeds) - 1`.

ilikecubesnstuff commented on utils/embeds.py:
```diff
@@ -0,0 +1,33 @@
+
+class EmbedGenerator():
+    def __init__(self, entries: List, title: Optional[str] = "", description: Optional[str] = "", step: Optional[int] = 10):
+       self.entries = entries
```
Personally, I like to add a line like `self.N = len(entries)` in my `__init__`s both for saving compute and brevity in the subsequent methods that might repeatedly require this value.

ilikecubesnstuff reviewed (commented):
I would add some docstrings inside Pagination and EmbedGenerator so its a bit clearer what they represent/what they do (though I might just be confused because this is the new part of discord's API).

NathanealV commented:
Commits should reflect the changes suggested by PR reviews!

ilikecubesnstuff commented on utils/commands.py:
```diff
@@ -2,12 +2,15 @@
-        except commands.CommandError as e:
-            logging.debug(f'Command "{cmd.name}" check threw error, discarded in {ctx.command.name} group subcommand list.')
+        except commands.CommandError:
+            logging.exception(f'Command "{cmd.name}" check threw error, discarded in {ctx.command.name} group subcommand list.')
```
> It might be valuable to log the error thrown as well here.

In hindsight, I wasn't the clearest here - I meant including the exception in the logged message, something like this:
```py
except commands.CommandError as e:
    logging.debug(f'Command "{cmd.name}" check threw error, discarded in {ctx.command.name} group subcommand list.\n\t{e}')
```
I think checks throwing exceptions is to be expected here so it's best to log at a `DEBUG` level, but logging the specific exception thrown could trace which check threw the error.

NathanealV replied:
What do you mean by "checks throwing exceptions is to be expected here"?

ilikecubesnstuff replied:
Like, when I was testing the updates cog, I remember discord.py's built-in permission checks throwing exceptions to signal a check failure.

NathanealV replied:
Hm so is the functionality provided by `logging.exception` not desired then? It prints the exception/stack trace as well [below specified text](https://stackoverflow.com/questions/5191830/how-do-i-log-a-python-error-with-debug-information).

ilikecubesnstuff replied:
Ah, I get it - that is a much better solution than the one in my comment. However, `logging.exception` logs at an `ERROR` level so it is still not quite the desired functionality - using `exc_info=True` inside `logging.debug` seems like a good idea (else this will log at an `ERROR` level every time a user lists subcommands and does not have permissions for all of them).

NathanealV replied:
Right good catch, should be changed now.

ilikecubesnstuff reviewed (commented):
The changes look good! I have 1 unresolved comment, but everything else looks great.

ilikecubesnstuff reviewed (commented):
LGTM!

raynichc commented on cogs/monitor.py:
```diff
@@ -13,96 +23,238 @@ async def on_message(self, message: discord.Message):
+        users_removed = 0
+
+        for monitor_user in monitor_users:
+            if not ctx.guild.get_member(monitor_user.user_id):
```
I'm not sure if its possible, but wouldn't it be better to check if the users had been banned as opposed to just not in the server anymore? I believe what is written is currently how its done, so I'm happy with it, but it may be better to change it. What do you think?

NathanealV replied:
That makes more sense to me yeah - should be added

raynichc commented on cogs/monitor.py:
```diff
@@ -13,96 +23,238 @@ async def on_message(self, message: discord.Message):
+        await ctx.send(f"{users_removed} user(s) were removed from monitor.")
     
     @monitor.command()
     async def list(self, ctx: commands.Context):
```
Not sure what would be the best way to go about it, but maybe having a way of juts viewing messages and just viewing users would be good.

raynichc commented on cogs/monitor.py:
```diff
@@ -13,96 +23,238 @@ async def on_message(self, message: discord.Message):
+    
+    @monitor.command()
+    @admin_command()
+    async def toggle(self, ctx: commands.Context):
```
I'm not sure if there is a use case for it, but we could maybe make it so that you can toggle on/off messages and user stuff separately?

NathanealV replied:
Not sure what the merit of this would be - this would also have to involve a migration of the DB schema

raynichc reviewed (commented):
I'm happy with the code, only comments are on potentially additions/modifications to the current flow. Not required, just thoughts, let me know what you think.

Only last thought is that we could potentially add monitor message groups, so that we can group messages that we want to monitor. That way when exams roll around, instead of having to add all the monitor tags back in, we could enable/disable an entire group of monitor messages at once. Let me know what you think, though this would be a pretty major change imo, so probably best to leave it off of this PR and push it to the future.

NathanealV commented:
The following suggestions will be implemented in a future PR from the same branch:
- Split list into message and users
- Split toggle into message and users
- Monitor groups
